xcvm/lib/core/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+mod abstraction;
+mod asset;
+mod instruction;
+mod network;
+mod program;
+mod protocol;
+
+use core::marker::PhantomData;
+
+pub use crate::{asset::*, instruction::*, network::*, program::*, protocol::*};
+use alloc::{collections::VecDeque, vec::Vec};
+
+#[inline]
+pub fn serialize_json<T: serde::Serialize>(
+	program: &Program<T>,
+) -> Result<Vec<u8>, serde_json::Error> {
+	serde_json::to_vec(program)
+}
+
+#[inline]
+pub fn deserialize_json<T: serde::de::DeserializeOwned>(
+	buffer: &[u8],
+) -> Result<Program<T>, serde_json::Error> {
+	serde_json::from_slice(buffer)
+}
+
+#[derive(Clone)]
+pub struct ProgramBuilder<CurrentNetwork: Network, Account, Assets> {
+	pub tag: Option<Vec<u8>>,
+	pub instructions: VecDeque<Instruction<NetworkID, Vec<u8>, Account, Assets>>,
+	pub _marker: PhantomData<CurrentNetwork>,
+}
+
+impl<CurrentNetwork, Account, Assets> ProgramBuilder<CurrentNetwork, Account, Assets>
+where
+	CurrentNetwork: Network,
+	CurrentNetwork::EncodedCall: Into<Vec<u8>>,
+{
+	#[inline]
+	pub fn new(tag: Option<Vec<u8>>) -> Self {
+		ProgramBuilder { tag, instructions: VecDeque::new(), _marker: PhantomData }
```

Comment:
afaik `ProgramBuilder { tag, ..Default::default() } could work, so not sure

Reply:
`ProgramBuilder` doesn't implement default, so it wouldn't work (you could do `		ProgramBuilder { tag, instructions: VecDeque::new(), _marker: Default::default() }` though, although that's probably not much of an improvement lol)